src/Pos/Block/Slog/Logic.hs

```diff
@@ -303,8 +307,16 @@ slogRollbackBlocks blunds = do
         mconcat [forwardLinksBatch, inMainBatch]
 
 -- Common actions for rollback and apply.
-slogCommon :: MonadSlogApply ssc ctx m => LastBlkSlots -> m ()
+slogCommon
+    :: MonadSlogApply ssc ctx m
+    => LastBlkSlots
+    -> m ()
 slogCommon newLastSlots = do
     sanityCheckDB
     slogPutLastSlots newLastSlots
-    putSlottingData =<< GS.getSlottingData
+    -- We read from the database and write in the memory.
+    -- TODO(ks): This is unsafe! We don't have control over sequentiality and
+    -- use explicit indexing. It would be better if we have sorted EpochSlotData and
+    -- pass it without the index.
```

Comment:
Well, if you agree to change `MonadSlotsData` to reader over `TVar SlottingData`, then this code can be simplified by using a single `writeTVar`.